Tests/ios 3261 add address tests for bip32

skibinalexander opened the pull request.

skibinalexander committed:
IOS-3261 Add any tests addresses with private key from mnemonic

skibinalexander committed:
IOS-3261 Add BIP test cases

skibinalexander committed:
IOS-3261 Add any tests address from public keys

skibinalexander committed:
IOS-3261 Refactor addresses compare tests

skibinalexander committed:
IOS-3261 Make for Mnemonic testcase

skibinalexander committed:
IOS-3261 Update init keys for testcases

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add any tests for BIP39 to Addresses blockchain

skibinalexander committed:
IOS-3261 Add Dash/Dogecoin testcases

skibinalexander committed:
IOS-3261 Add any testcases

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Refactor tests utility

skibinalexander committed:
IOS-3261, IOS-3349, IOS-3348 Completed public BIP32 from tangem-sdk

tureck1y commented on BlockchainSdkTests/Common/BIP39ServiceManagerUtility.swift:
```diff
@@ -0,0 +1,139 @@
+
+@testable import BlockchainSdk
+
+final class MnemonicServiceManagerUtility {
```
название отличается от 
// BIP39ServiceManagerUtility.swift
и от названия файла

skibinalexander replied:
Ага, пропустил ...

tureck1y reviewed (commented):
Немного не хватает комментов, чтобы понять где что тестируется, откуда и как создаются ключи, адреса. Что с чем конкретно проверяется в конкретном файле, вот.

skibinalexander committed:
IOS-3261 Rename filename MnemonicServiceManagerUtility

skibinalexander committed:
IOS-3261 Refactor name MnemonicServiceManagerUtility.swift

skibinalexander commented:
> Немного не хватает комментов, чтобы понять где что тестируется, откуда и как создаются ключи, адреса. Что с чем конкретно проверяется в конкретном файле, вот.

Да ок, сделаю

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj

skibinalexander committed:
IOS-3261 Add documentation for use Common tests

skibinalexander commented:
> Немного не хватает комментов, чтобы понять где что тестируется, откуда и как создаются ключи, адреса. Что с чем конкретно проверяется в конкретном файле, вот.

@tureck1y погляди пожалуйста, как время будет, добавил доку

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	BlockchainSdk/Extensions/TWCoin+.swift

skibinalexander committed:
IOS-3261 Add TWCoin+ for ravencoin

skibinalexander committed:
IOS-3261 Refactor Coin+

tureck1y reviewed (commented):
На мой взгляд сложновато получилось, думал комменты помогут.
1) есть несколько классов с тестами, в которых тестируются адреса. Почему один тест в одном файле, а другой в другом - непонятно при беглом просмотре. 

2) очень много утилит и разных конвертаций. Со стороны непонятен путь, который проходят ключи. Нужно идти по цепочкам вглубь утилит и разбираться. 

Мне как разработчику нового блокчейна придется долго разбираться куда добавить тест моего нового блокчейна. Я ожидаю увидеть полотно одинаковых тестов в одном файле, по аналогии с тестами на генерацию адреса. Пускай там будет копипаста, но мне будет понятно. Вот мнемоника. Тут она превратилась в адрес в трастволлете, а тут она превратилась в адрес в нашем СДК. 

Как я вижу эти тесты:
1) Берем мнемонику, загоняем ее на карту делаем все необходимые деривации, получаем ключи
2) по каждому блокчейну дергаем AddressService с соотв. ключом (нужная кривая и деривация) и получаем наш адрес
3) превращаем мнемонику в адрес исключительно средствами трастволлета для нужного блокчейна
4) сравниваем адреса

Если у тв в тестах уже есть мнемоника -> адрес, можно просто взять уже готовые адреса и сравнивать с нашими, а не генерировать.

skibinalexander commented:
> На мой взгляд сложновато получилось, думал комменты помогут.
> 
> 1. есть несколько классов с тестами, в которых тестируются адреса. Почему один тест в одном файле, а другой в другом - непонятно при беглом просмотре.
> 2. очень много утилит и разных конвертаций. Со стороны непонятен путь, который проходят ключи. Нужно идти по цепочкам вглубь утилит и разбираться.
> 
> Мне как разработчику нового блокчейна придется долго разбираться куда добавить тест моего нового блокчейна. Я ожидаю увидеть полотно одинаковых тестов в одном файле, по аналогии с тестами на генерацию адреса. Пускай там будет копипаста, но мне будет понятно. Вот мнемоника. Тут она превратилась в адрес в трастволлете, а тут она превратилась в адрес в нашем СДК.
> 
> Как я вижу эти тесты:
> 
> 1. Берем мнемонику, загоняем ее на карту делаем все необходимые деривации, получаем ключи
> 2. по каждому блокчейну дергаем AddressService с соотв. ключом (нужная кривая и деривация) и получаем наш адрес
> 3. превращаем мнемонику в адрес исключительно средствами трастволлета для нужного блокчейна
> 4. сравниваем адреса
> 
> Если у тв в тестах уже есть мнемоника -> адрес, можно просто взять уже готовые адреса и сравнивать с нашими, а не генерировать.

Понял, сегодня подумаю.

skibinalexander committed:
IOS-3261 Refactor testcases for any

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Remove methods for any files

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add validation derivationPath for any blockchain

skibinalexander committed:
IOS-3261 Add any tests for use derivation

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add validation publicKey from tangemSdk

skibinalexander committed:
IOS-3261 Add any cases public keys from tangemSdk

skibinalexander committed:
IOS-3261 Add Avalanch public key

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add implementation tests derivations for use after review

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	BlockchainSdk/Extensions/TWCoin+.swift
#	Podfile.lock

tureck1y commented on BlockchainSdkTests/Common/BlockchainServiceManagerUtility.swift:
```diff
@@ -0,0 +1,213 @@
+    ]
+    
+    let sdkDerivations: [DerivationUnion] = [
+        .init(path: "m/44'/0'/0'/0/0", blockchain: .bitcoin(testnet: false)),
```
а зачем так? можно же прям из enam вытащить

tureck1y commented on BlockchainSdkTests/Common/AddressServiceManagerUtility.swift:
```diff
@@ -15,25 +15,14 @@ import WalletCore
-        }
-    }
-    
     func validate(address: String, publicKey: Data, for blockchain: BlockchainSdk.Blockchain) {
```
осталась вариация в утилите

tureck1y commented on BlockchainSdkTests/Common/BlockchainServiceManagerUtility.swift:
```diff
@@ -0,0 +1,213 @@
+    ]
+    
+    let twDerivations: [DerivationUnion] = [
+        .init(path: "m/84'/0'/0'/0/0", blockchain: .bitcoin(testnet: false)),
```
это из registry.json? Надо бы комемнтик тогда

tureck1y commented on BlockchainSdkTests/Common/BlockchainServiceManagerUtility.swift:
```diff
@@ -0,0 +1,213 @@
+        .init(key: "023A291DBF3427922A22D166C8D3333253A0E383115CF5DB36BE47EC5EA81D7D56", blockchain: .kava(testnet: false)),
+        .init(key: "035F25F3544EF2D031DB235B341112C31F1B26C03F13DC52B5D0E2D508A4F3F176", blockchain: .kaspa),
+        .init(key: "0260BF54E0D66880E36FC5400C5CBF799AE129807B03B0FC88DB2506217EB24D85", blockchain: .ravencoin(testnet: false)),
+    ]
```
мне кажется нужно составить тестовый вектор просто со всеми данными, например такой json
{
""
mnemonic
curve
key
address
...
""
}

Будет массив с такими кусочками данных и по ним уже тесты гнать, наподобие тестов на BIP39 в карточном сдк

tureck1y commented on BlockchainSdkTests/CompareTests/DerivationCompareTests.swift:
```diff
@@ -0,0 +1,76 @@
+    
+    // MARK: - Implementation
+    
+    func testDerivationsFromBlockchainPath() {
```
я не совсем понял этот метод.
Есть деривация от TW
Есть деривация от нас
а что такое sdkreference?

tureck1y reviewed (commented):
В целом мы на правильном пути. Я думаю стоит добавить тест вектора и тогда сравниваться все будет в одном месте, в одном файле с тестами, завтра обсудим еще

skibinalexander committed:
IOS-3261 Fix after review comments

skibinalexander committed:
IOS-3261 Add any vectors for validate addresses from public key

skibinalexander committed:
IOS-3261 Refactor name files

skibinalexander committed:
IOS-3261 add mnemonics and refactoring

skibinalexander committed:
IOS-3261 Add vector suggestions

skibinalexander committed:
IOS-3261 Add version master keys

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add implementation private keys tests

skibinalexander committed:
IOS-3261 Revert changes main tests

skibinalexander committed:
IOS-3261 Refactor tests bitcoin

skibinalexander committed:
IOS-3261 fix bitcoin bad_access

skibinalexander committed:
IOS-3261 Refactor unuse files

skibinalexander committed:
IOS-3261 IOS-3520 Add implementation vectors from trustwallet

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander commented:
@tureck1y Посмотри пожалуйста, пока не мерджить тесты валятся.

skibinalexander commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -19,6 +19,7 @@ class AddressesTests: XCTestCase {
     private let secpDecompressedKey = Data(hexString: "0441DCD64B5F4A039FC339A16300A833A883B218909F2EBCAF3906651C76842C45E3D67E8D2947E6FEE8B62D3D3B6A4D5F212DA23E478DD69A2C6CCC851F300D80")
```
Тут сегодня еще докину

skibinalexander commented on BlockchainSdkTests/Stellar/StellarTests.swift:
```diff
@@ -24,17 +24,6 @@ class StellarTests: XCTestCase {
         bag = []
     }
     
-    func testAddress() {
```
тут сделаю реверс , случайно попало

skibinalexander commented on BlockchainSdkTests/Litecoin/LitecoinTests.swift:
```diff
@@ -19,17 +19,6 @@ class LitecoinTests: XCTestCase {
     
     private lazy var addressService = blockchain.getAddressService()
     
-    func testAddress() {
```
сделаю реверт случайно попало

skibinalexander committed:
IOS-3261 Remove changes from tests

skibinalexander committed:
IOS-3261 Add test any test for validation with trust wallet

tureck1y commented on BlockchainSdkTests/Common/AddressServiceManagerUtility.swift:
```diff
@@ -15,60 +15,33 @@ import WalletCore
     }
     
-    private func makeLocalWalletAddressService(
+    func makeLocalWalletAddressService(
```
makeTangemAddressService?

Правда создается адрес, а не AddressService..

makeTangemAddress\makeTrustWalletAddress ?

tureck1y commented on BlockchainSdkTests/Common/DecodableVectors.swift:
```diff
@@ -0,0 +1,46 @@
+    case blockchain = "blockchain_vectors"
+    case validAddress = "valid_address_vectors"
+    case derivation = "derivation_vectors"
+    case turstWalletCompare = "trust_wallet_comapre_vector"
```
turst

tureck1y reviewed (commented):
В целом вроде ок, можно уже открывать и закомментить failed тесты с TODO, чтобы их удобно было открыть, когда настанет время

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

skibinalexander committed:
IOS-3261 Add wallet 2.0 public key vector

skibinalexander committed:
IOS-3261 Rename methods

skibinalexander committed:
IOS-3261 Make test completed

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	Podfile.lock

Andoran90 commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -0,0 +1,115 @@
+
+extension PublicKeyDerivationAddressTests {
+    
+    func testPublicKeyDerivationAddressVector() {
```
могу ошибаться, но думаю что тут слишком дофига всего тестируется. Все же юнит тесты предусматривают более точечное тестирование. Например тестировать создание приватного ключа из сид фразы, тестирование создания публичного ключа без деривации, тестирование создания публичного ключа с деривацией, тестирование создания адреса из публичного ключа с деривацией. Например блокчеин вообще никак не влияет на создание приватного ключа, семечка для BIP32 тоже никак не влияет, а вот публичный ключ с деривацией уже имеет смысл. 
Мне кажется лучше много маленьких точечных тестов, чем один огромный тест на все случаи жизни. Что скажете, @skibinalexander @tureck1y ?
Почему говорю так: приватный ключ и сид вообще не относятся к этому тесту, по идее их тут даже проверять не надо, это должен быть 2 отдельных теста, один на получение семечки и второй на генерацию приватного ключа и привязаны быть к кривым, а не к блокчеинам. Причем во втором не надо проверять правильность семечки, т.к. она была уже протестирована в первом тесте. А тут получается 17 тестов и в каждом проверяется правильно ли создались приватные ключи.

tureck1y replied:
Да, я согласен, что лучше много мелких. Тут я предлагал сделать тестирование по аналогии с тест-векторами в разных BIPах. Чтобы в векторе был прописать путь от мнемоники до адреса, например и при тестировании вектора, тестировались все этапы и падали точечно.

Добавили новый блокчейн - добавили вектор, тесты писать не нужно дополнительно. Тогда не будет ситуации, что забыли что-то протестировать при добавлении нового блокчейна.

Andoran90 replied:
но при добавлении нового блокчеина надо же проверить - что публичный ключ деривируется правильно и адрес создается верно, тогда и тесты не надо писать дополнительные...

Но если план сделать тесты по аналогии как и в бипах, ну оукей))

Andoran90 commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -0,0 +1,115 @@
+ - 0. Compare derivation from vector file with BlockchainSdk.derivationPath(.new)
+ - 1. Obtain MASTER Trustwallet Keys and keys from TrangemSdk and compare keys
+ - 2. Obtain PUBLIC Trustwallet Keys and keys from TrangemSdk and compare keys
+ - 2. Obtain ADDRESSES TeustWallet service and BlockchainSdk service for derivation public keys
```
```suggestion
 - 3. Obtain ADDRESSES TrustWallet service and BlockchainSdk service for derivation public keys
```

а в коде получаем только публичный ключ из траста

skibinalexander replied:
Поправил описание

Andoran90 commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -0,0 +1,115 @@
+                
+                guard CoinType(blockchain) != nil else { return }
+                
+                //FIXME: - Remove when all test will be is completed
+                guard !(test.skip ?? false) else { return }
```
думаю эту проверку надо передвинуть в начало кложуры и выполнять её вперед всего

megakoko replied:
А почему поле skip опциональное?

Andoran90 replied:
может чтобы json был поменьше и легче было выцепить какой вектор отключен?..

skibinalexander replied:
> может чтобы json был поменьше и легче было выцепить какой вектор отключен?..

В виду того, что есть расхождения по ключам и деривациям, тесты падают, добавили скип, как исправим в блокчуейн сок деривации и корректность адресов, скип убрать

Andoran90 commented on BlockchainSdkTests/VectorTests/AddressesValidationTests.swift:
```diff
@@ -0,0 +1,60 @@
+                
+                vector.negative.forEach {
+                    XCTAssertFalse(TrustWalletAddressService.validate($0, for: blockchain), "-> \(blockchain)")
+                    XCTAssertFalse(blockchain.getAddressService().validate($0), "-> \(blockchain)")
```
а тут почему-то ни в одном нет `__INVALID_VECTOR__`... и никакого толком сообщения об ошибке. 

В итоге получается вот такая запись:
blockchain-sdk-swift/BlockchainSdkTests/VectorTests/AddressesValidationTests.swift:51: error: -[BlockchainSdkTests.AddressesValidationTests testAddressVector] : XCTAssertFalse failed - -> polygon(testnet: false)

Может лучше написать что именно не так без всяких векторов инвалидов? В любом случае если тест провалился значит где-то засел инвалид, лучше указать причину.

skibinalexander replied:
__INVALIDATE_SDK_CORE__ добавил группу ошибки чтобы повысить читаемость

megakoko commented on BlockchainSdkTests/Common/KeysServiceManagerUtility.swift:
```diff
@@ -0,0 +1,82 @@
+    
+    // MARK: - Properties
+    
+    private var mnemonic: String = ""
```
И значение по-умолчанию убрать, только смущает

megakoko commented on BlockchainSdkTests/Common/KeysServiceManagerUtility.swift:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+    
+    // MARK: - asjkdaksldlk
```
asjkdaksldlk!

Andoran90 replied:
rjfkrepbmrekju?

Andoran90 commented:
а ещё надо поменять названия на `WalletCore`)

megakoko commented on BlockchainSdkTests/Vectors/trust_wallet_comapre_vector.json:
..._compare_vector...

skibinalexander committed:
IOS-3261 Merge branch 'develop' into tests/IOS-3261-Add-address-tests-for-bip32

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdk/Extensions/TWCoin+.swift
#	BlockchainSdkTests/Common/AddressServiceManagerUtility.swift

skibinalexander committed:
IOS-3261

skibinalexander committed:
IOS-3261 Update review comments

skibinalexander committed:
IOS-3261 Update documentation

skibinalexander committed:
IOS-3261 Add implementation comments for tests

skibinalexander commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -0,0 +1,115 @@
+            
+            try vector.testable.forEach { test in
+                guard let blockchain = blockchains.first(where: { $0.codingKey == test.blockchain }) else {
+                    print("__INVALID_VECTOR__ MATCH BLOCKCHAIN KEY IS NIL \(test.blockchain)")
```
Могу убрать, но лучше в качестве лога выводить

skibinalexander replied:
> почему тут просто печатается в консоль ошибка? она должна отображаться как провалившийся тест

Сигнализирует что маппинг вектора не правильный

Andoran90 replied:
я не понимаю почему просто в консоль печатать? если не удалось получить блокчеин, а тест вектор есть - значит тест провален: либо не хватает блокчеина, либо тествектор лишний, разве нет?

skibinalexander commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -0,0 +1,115 @@
+                // MARK: -  Step - 1 / 2
+                
+                let keysServiceUtility = KeysServiceManagerUtility(mnemonic: vector.mnemonic.words)
+                let seed = try keysServiceUtility.getTrustWalletSeed()
```
> не понимаю смысла этого. Сейчас создается утилита, которая имеет стейт, помимо этого эта утилита хранит мнемонику и сама генерирует новые сущности `HDWallet`. А почему утилита не может принять на вход информацию о блокчеине, сразу создать все необходимые приватные, публичные ключи? Сейчас получается что туда-сюда катается сид, но он не тестируется, правильная ли дата создалась. Интерфейс утилиты усложняется и без особого смысла для тестов. 
> 
> 
> 
> Я так понимаю что лучше чтобы утилита не хранила никакой стейт, а просто брала на вход данные, делала с нею магию и выдавала результат. Или же чтобы она инициализировалась, создавала все нужные сущности (пары ключей) и выдавала, что уже нагенерировала.

Стейт хранится из за особенностей методов траст волета, предлагаю пока оставить , ещё скорее всего будут изменения там и придумаю что нибудь получше

Andoran90 replied:
а какие особенности? надо обязательно каждый раз пересоздавать HDWallet что ли? или что?

skibinalexander commented on BlockchainSdkTests/Common/KeysServiceManagerUtility.swift:
```diff
@@ -0,0 +1,82 @@
+        if let coin = CoinType(blockchain) {
+            return try hdWallet.getKey(coin: coin, derivationPath: derivation).getPublicKeyByType(pubkeyType: .init(blockchain))
+        } else {
+            throw NSError(domain: "__INVALID_EXECUTE_KEY__ BLOCKCHAIN \(blockchain.currencySymbol)", code: -1)
```
> выбрасывание этой ошибки и в предыдущей функции ничем не отличаются, получается если один тест упадет будет не особо информативно. Думаю можно сделать через `#function` или же изменить текст `domain`. Думаю второй вариант предпочтительней

Предложить надо будет какой текст предпочтительней )

skibinalexander commented on BlockchainSdkTests/VectorTests/AddressesValidationTests.swift:
```diff
@@ -0,0 +1,60 @@
+    func testAddressVector() {
+        do {
+            guard let blockchains: [BlockchainSdk.Blockchain] = try testVectorsUtility.getTestVectors(from: "blockchain_vectors") else {
+                XCTFail("__INVALID_VECTOR__ BLOCKCHAIN DATA IS NIL")
```
Поменял текста

skibinalexander commented:
@Andoran90 Задача в том чисом ле как раз таки стояла в том чтобы за счет того, что мы в целом ориентируемся на тесты траст волета при добавлении блокчейна, автоматизировать за счет векторов процесс тестирования нового блокчейна.

Текстовки __ GROUP ERROR __ в целом для того, чтобы выделить корень ошибки и сделать более читаемой

print поиска блокчейна заменил XCFail

skibinalexander committed:
IOS-3261 Fix review comments

skibinalexander committed:
IOS-3261 Fix review comments
